src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -35,92 +35,49 @@ private CSharpSyntaxNode CheckRecursivePatternFeature(CSharpSyntaxNode node)
 
         private CSharpSyntaxNode ParseTypeOrPatternForIsOperatorCore()
         {
-            var tk = this.CurrentToken.Kind;
-            Precedence precedence = GetPrecedence(SyntaxKind.IsPatternExpression);
-
-            // We will parse a shift-expression ONLY (nothing looser) - i.e. not a relational expression
-            // So x is y < z should be parsed as (x is y) < z
-            // But x is y << z should be parsed as x is (y << z)
-            Debug.Assert(Precedence.Shift == precedence + 1);
-
-            // For totally broken syntax, parse a type for error recovery purposes
-            switch (tk)
+            var pattern = ParsePattern(GetPrecedence(SyntaxKind.IsPatternExpression), afterIs: true);
+            return pattern switch
             {
-                case SyntaxKind.IdentifierToken when this.CurrentToken.ContextualKind == SyntaxKind.UnderscoreToken:
-                // We permit a type named `_` on the right-hand-side of an is operator, but not inside of a pattern.
-                case SyntaxKind.CloseParenToken:
-                case SyntaxKind.CloseBracketToken:
-                case SyntaxKind.CloseBraceToken:
-                case SyntaxKind.SemicolonToken:
-                case SyntaxKind.CommaToken:
-                    // HACK: for error recovery, we prefer a (missing) type.
-                    return this.ParseType(ParseTypeMode.AfterIs);
-                default:
-                    // attempt to disambiguate.
-                    break;
-            }
+                ConstantPatternSyntax cp when ConvertExpressionToType(cp.Expression, out NameSyntax type) => type,
+                TypePatternSyntax tp => tp.Type,
+                DiscardPatternSyntax dp => _syntaxFactory.IdentifierName(ConvertToIdentifier(dp.UnderscoreToken)),
+                var p => p,
```

Comment:
`_ => pattern`?

Reply:
I prefer this form, so that the result is entirely a function of the already computed input.

---
In reply to: [354971890](https://github.com/dotnet/roslyn/pull/39913#discussion_r354971890) [](ancestors = 354971890)